sycl/doc/CompileTimeProperties.md

```diff
@@ -0,0 +1,680 @@
+    Props::meta_name..., Props::meta_value...
+    )]]
+#endif
+  accessor<dataT,
+           dimensions,
+           accessmode,
+           accessTarget,
+           isPlaceholder,
+           property_list<Props...>> {/*...*/};
+
+} // namespace sycl
+```
+
+Illustrating this with the previous example:
+
+```
+namespace sycl {
+
+template </* ... */> class
+#ifdef __SYCL_DEVICE_ONLY__
+  [[__sycl_detail__::add_ir_kernel_parameter_attributes(
+    "sycl-no-alias",  // Name of first property
+    "sycl-foo",       // Name of second property
+    "",               // First property has no parameter
+    32                // Value of second property
+    )]]
+#endif
+  accessor</* ... */> {/* ... */};
+
+} // namespace sycl
+```
+
+As the name of the C++ attribute suggests, the device compiler front-end uses
+the attribute only when the decorated type is the type of a kernel argument,
+and it silently ignores the attribute when the class is used in any other way.
+
+When the device compiler front-end creates a kernel argument in this way, it
+adds one LLVM IR attribute to the kernel function's parameter for each property
+in the list.  For example, this can be done by calling
+[`Function::addParamAttrs(unsigned ArgNo, const AttrBuilder &)`][7].  As
+before, the IR attributes are added as strings, so the front-end must convert
+the property value to a string if it is not already a string.
+
+[7]: <https://llvm.org/doxygen/classllvm_1_1Function.html#a092beb46ecce99e6b39628ee92ccd95a>
+
+**TODO**: What happens when a "sycl special class" object is captured as a
+kernel argument?  The compiler passes each member of the class as a separate
+argument.  Should the device compiler duplicate the properties on each such
+parameter in this case?  Or, is it the header's responsibility to add the C++
+attribute to one of the member variables in this case?  How does the header
+decide which member variable to decorate, though?
+
+
+## Properties on kernel functions
+
+Compile-time properties can also be used to decorate kernel functions as with
+the [SYCL\_EXT\_ONEAPI\_KERNEL\_PROPERTIES][8] extension.  There are two ways
+the application can specify these properties.  The first is by passing a
+`property_list` parameter to the function that submits the kernel:
+
+[8]: <https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/KernelProperties/KernelProperties.asciidoc>
+
+```
+namespace sycl {
+
+class handler {
+  template <typename KernelName, typename KernelType, typename PropertyListT>
+  void single_task(PropertyListT properties, const KernelType &kernelFunc);
+};
+
+// namespace sycl
+```
+
+For example:
+
+```
+using sycl;
+using sycl::ext::oneapi;
+
+void foo(handler &cgh) {
+  cgh.single_task(
+    property_list{sub_group_size_v<32>, device_has_v<aspect::fp16>},
+    [=] {/* ... */});
+}
+```
+
+The second way an application can specify kernel properties is by adding a
+`properties` member variable to a named kernel function object:
```

Comment:
This name can collide standard SYCL2020 code. Can we rename it to something like `_properties`?
Also, it's expected that there are some checks in the headers verifying that this var is of expected type, right?

Reply:
I don't see how the name "properties" can collide with anything defined in the core SYCL 2020 spec.  Can you elaborate?  It so happens that we are expecting to update the SYCL_EXT_ONEAPI_PROPERTY_LIST extension spec, though, to define a type named "properties".  When this happens, we will also rename this member variable to something else to avoid confusion.  @Pennycook is working on a proposal for this.

Yes, the headers should probably have a `static_assert` that the type of the `properties` member variable is `sycl::ext::oneapi::property_list`.

Reply:
@romanovvlad Is the concern that the user might already have a variable called "properties" in their functor (and nothing would stop them from doing so) because "properties" isn't a keyword?  I don't really like the idea of making it a minimizing conflicts by adding an underscore -- users won't understand why the _ is necessary, and I'd expect them to complain.

We're considering a few alternative designs that wouldn't rely on a member variable at all, but we're having some issues coming up with something that's easy to use and isn't too verbose.  If you have any suggestions for alternative ways to attach a property list to a functor, please let me know.  We've not yet committed to the named member variable solution.

Reply:
>> s the concern that the user might already have a variable called "properties" in their functor (and nothing would stop them from doing so) because "properties" isn't a keyword? 

Exactly.

>> If you have any suggestions for alternative ways to attach a property list to a functor, please let me know. 

If adding underscores is no-go solution. Then can suggest thinking about asking users to inherit their functor from a SYCL defined class which will add a variable with underscores implicitly, something like:

```

class MyKernel: public sycl::ext_one_api_propertize<sub_group_size_v<32>, device_has_v<aspect::fp16>> {
  public:
    void operator()() {/* ... */}
};
```
```
template<class... T>
class ext_one_api_propertize {
    static constexpr auto __properties = property_list{T...}; 
};

Reply:
I'm marking this as resolved because the mechanism for declaring properties of a named function object is not central to this design doc.  @Pennycook is working on an API extension specification for this, and his document will address these issues.  When the API is finalized, we can update this design document.